src/test/aztec/base/BridgeTestBase.sol

```diff
@@ -0,0 +1,407 @@
+        uint256 indexed bridgeId,
+        uint256 indexed nonce,
+        uint256 totalInputValue,
+        uint256 totalOutputValueA,
+        uint256 totalOutputValueB,
+        bool result,
+        bytes errorReason
+    );
+    event AsyncDefiBridgeProcessed(uint256 indexed bridgeId, uint256 indexed nonce, uint256 totalInputValue);
+    event Deposit(uint256 indexed assetId, address indexed depositorAddress, uint256 depositValue);
+    event WithdrawError(bytes errorReason);
+    event AssetAdded(uint256 indexed assetId, address indexed assetAddress, uint256 assetGasLimit);
+    event BridgeAdded(uint256 indexed bridgeAddressId, address indexed bridgeAddress, uint256 bridgeGasLimit);
+    event RollupProviderUpdated(address indexed providerAddress, bool valid);
+    event VerifierUpdated(address indexed verifierAddress);
+    event Paused(address account);
+    event Unpaused(address account);
+
+    uint256 private constant INPUT_ASSET_ID_A_SHIFT = 32;
+    uint256 private constant INPUT_ASSET_ID_B_SHIFT = 62;
+    uint256 private constant OUTPUT_ASSET_ID_A_SHIFT = 92;
+    uint256 private constant OUTPUT_ASSET_ID_B_SHIFT = 122;
+    uint256 private constant BITCONFIG_SHIFT = 152;
+    uint256 private constant AUX_DATA_SHIFT = 184;
+    uint256 private constant VIRTUAL_ASSET_ID_FLAG_SHIFT = 29;
+    uint256 private constant VIRTUAL_ASSET_ID_FLAG = 0x20000000; // 2 ** 29
+    uint256 private constant MASK_THIRTY_TWO_BITS = 0xffffffff;
+    uint256 private constant MASK_THIRTY_BITS = 0x3fffffff;
+    uint256 private constant MASK_SIXTY_FOUR_BITS = 0xffffffffffffffff;
+
+    IRollupProcessor internal constant ROLLUP_PROCESSOR = IRollupProcessor(0xFF1F2B4ADb9dF6FC8eAFecDcbF96A2B351680455);
+    IRollupProcessor internal constant IMPLEMENTATION = IRollupProcessor(0x3f972e325CecD99a6be267fd36ceB46DCa7C3F28);
+
+    address internal constant ROLLUP_PROVIDER = payable(0xA173BDdF4953C1E8be2cA0695CFc07502Ff3B1e7);
+    address internal constant MULTI_SIG = 0xE298a76986336686CC3566469e3520d23D1a8aaD;
+
+    AztecTypes.AztecAsset internal emptyAsset;
+
+    uint256 private nextRollupId = 0;
+
+    constructor() {
+        vm.label(address(ROLLUP_PROCESSOR), "Rollup");
+        vm.label(address(IMPLEMENTATION), "Implementation");
+        vm.label(ROLLUP_PROVIDER, "Rollup Provider");
+        vm.label(MULTI_SIG, "Multisig");
+        vm.label(ROLLUP_PROCESSOR.verifier(), "Verifier");
+        vm.label(ROLLUP_PROCESSOR.defiBridgeProxy(), "DefiBridgeProxy");
+    }
+
+    function getNextNonce() public view returns (uint256) {
+        return nextRollupId * 32;
+    }
+
+    function checkValidAsset(AztecTypes.AztecAsset memory _asset) public {
+        address asset = ROLLUP_PROCESSOR.getSupportedAsset(_asset.id);
+        assertEq(asset, _asset.erc20Address, "ERC20 address do not match");
+        // TODO: Finish
+    }
+
+    function getRealAztecAsset(address _asset) public view returns (AztecTypes.AztecAsset memory res) {
```

Comment:
Ideally I would add Natspec docs everywhere in the bridge base so that grantees are much less confused. The goal should be that grantees can gain a sufficient understanding from ExampleBridge and ExampleBridgeTest to write a good bridge.